homeassistant/components/swidget/__init__.py

```diff
@@ -0,0 +1,108 @@
+"""The swidget integration."""
+
+from datetime import timedelta
+import logging
+from typing import Any
+
+from swidget.discovery import SwidgetDiscoveredDevice, discover_devices, discover_single
+from swidget.swidgetdevice import SwidgetDevice
+
+from homeassistant.config_entries import SOURCE_INTEGRATION_DISCOVERY, ConfigEntry
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_MAC,
+    CONF_NAME,
+    CONF_PASSWORD,
+    EVENT_HOMEASSISTANT_STARTED,
+    Platform,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DOMAIN
+from .coordinator import SwidgetDataUpdateCoordinator
+
+LOGGER = logging.getLogger(__name__)
+PLATFORMS: list[Platform] = [Platform.LIGHT]
+DISCOVERY_INTERVAL = timedelta(minutes=15)
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+
+SwidgetConfigEntry = ConfigEntry[SwidgetDataUpdateCoordinator]
+
+
+@callback
+def async_trigger_discovery(
+    hass: HomeAssistant,
+    discovered_devices: dict[str, SwidgetDiscoveredDevice],
+) -> None:
+    """Trigger config flows for discovered devices."""
+    for device in discovered_devices.values():
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_INTEGRATION_DISCOVERY},
+                data={
+                    CONF_NAME: device.friendly_name,
+                    CONF_HOST: device.host,
+                    CONF_MAC: device.mac,
+                },
+            )
+        )
+
+
+async def async_discover_devices(
+    hass: HomeAssistant,
+) -> dict[str, SwidgetDiscoveredDevice]:
+    """Force discover Swidget devices using SSDP."""
+    devices: dict[str, SwidgetDiscoveredDevice]
+    devices = await discover_devices(timeout=15)  # type: ignore [no-untyped-call]
+    return devices
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Swidget component."""
+    hass.data[DOMAIN] = {}
+    if discovered_devices := await async_discover_devices(hass):
+        async_trigger_discovery(hass, discovered_devices)
+
+    async def _async_discovery(*_: Any) -> None:
+        if discovered := await async_discover_devices(hass):
+            async_trigger_discovery(hass, discovered)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, _async_discovery)
+    async_track_time_interval(hass, _async_discovery, DISCOVERY_INTERVAL)
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: SwidgetConfigEntry) -> bool:
+    """Set up swidget from a config entry."""
+
+    device = await discover_single(
+        host=entry.data[CONF_HOST],
+        token_name="x-secret-key",
+        password=entry.data[CONF_PASSWORD],
+        use_https=True,
+        use_websockets=True,
+    )
```

Comment:
We don't check the type of device returned here, or pass the type we expect, yet we're confident enough to cast from `SwidgetDevice` to `SwidgetDimmer` when setting up the light. I think we need to check we get the expected device somewhere?

Reply:
Updated. Please check